### PR TITLE
builtin_macros: Fix use of interpolated identifiers in `asm!`

### LINKED ISSUE #77584: macro_rules-wrapped asm! fails on att_syntax
The [`probe`](https://github.com/cuviper/rust-libprobe) crate uses `asm!` with `att_syntax`, because SDT operands are written in ELF notes, to be read by third-party tools expecting that syntax. Recently this has [regressed](https://github.com/cuviper/rust-libprobe/runs/1204132986?check_suite_focus=true):

```
~/rust/probe$ cargo check --examples
    Checking probe v0.2.1 (/home/jistone/rust/probe)
error: expected one of `)`, `att_syntax`, `nomem`, `noreturn`, `nostack`, `preserves_flags`, `pure`, or `readonly`, found `att_syntax`
 --> examples/loop.rs:4:5
  |
4 |     probe!(foo, begin);
  |     ^^^^^^^^^^^^^^^^^^^ expected one of 8 possible tokens
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected one of `)`, `att_syntax`, `nomem`, `noreturn`, `nostack`, `preserves_flags`, `pure`, or `readonly`, found `att_syntax`
 --> examples/loop.rs:8:9
  |
8 |         probe!(foo, loop, i, total);
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected one of 8 possible tokens
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected one of `)`, `att_syntax`, `nomem`, `noreturn`, `nostack`, `preserves_flags`, `pure`, or `readonly`, found `att_syntax`
  --> examples/loop.rs:11:5
   |
11 |     probe!(foo, end);
   |     ^^^^^^^^^^^^^^^^^ expected one of 8 possible tokens
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 3 previous errors

error: could not compile `probe`

To learn more, run the command again with --verbose.
```

`cargo bisect-rustc` pointed me to #77275 -- cc @petrochenkov. I see that there were regressions with `asm!` during that review, but not exactly like what I'm seeing here.

### Meta

searched nightlies: from nightly-2020-09-28 to nightly-2020-10-04
regressed nightly: nightly-2020-09-30
searched commits: from https://github.com/rust-lang/rust/commit/fc2daaae610b5515438b551a2f3706196a997f35 to https://github.com/rust-lang/rust/commit/381b445ff5751f9f39ec672b623372dff09c276e
regressed commit: https://github.com/rust-lang/rust/commit/26373fb4baa9c5b8a7a1e2821fcfa930a85d327d

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.5.2</summary>

Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --access github -- check --examples
```
</details>

### PR DESCRIPTION
**Pull request by petrochenkov**
Fixes https://github.com/rust-lang/rust/issues/77584

### CODE REVIEW

File: compiler/rustc_builtin_macros/src/asm.rs [hunk 1]
```diff
@@ -333,21 +333,22 @@ fn parse_options<'a>(p: &mut Parser<'a>, args: &mut AsmArgs) -> Result<(), Diagn
...
             try_set_option(p, args, sym::noreturn, ast::InlineAsmOptions::NORETURN);
-        } else if p.eat(&token::Ident(sym::nostack, false)) {
+        } else if p.eat_keyword(sym::nostack) {
             try_set_option(p, args, sym::nostack, ast::InlineAsmOptions::NOSTACK);
-        } else {
-            p.expect(&token::Ident(sym::att_syntax, false))?;
+        } else if p.eat_keyword(sym::att_syntax) {
             try_set_option(p, args, sym::att_syntax, ast::InlineAsmOptions::ATT_SYNTAX);
+        } else {
+            return Err(p.expect_one_of(&[], &[]).unwrap_err());
```
Comment (oli-obk) [inline compiler/rustc_builtin_macros/src/asm.rs]: shouldn't we specify the full list of things that we could consume here?

File: src/test/ui/asm/interpolated-idents.stderr [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error: asm outputs are not allowed with the `noreturn` option
+  --> $DIR/interpolated-idents.rs:10:32
+   |
+LL |               asm!("", $in(x) x, $out(x) x, $lateout(x) x, $inout(x) x, $inlateout(x) x,
+   |                                  ^^^^^^^^^  ^^^^^^^^^^^^^  ^^^^^^^^^^^  ^^^^^^^^^^^^^^^
+...
+LL |       m!(in out lateout inout inlateout const sym
+   |  _____-
```
Comment (oli-obk) [inline src/test/ui/asm/interpolated-idents.stderr:34]: oh wow... cc @rust-lang/wg-diagnostics we should probably deduplicate `span_label`s that have the same message and span

Comment (Aaron1011) [reply src/test/ui/asm/interpolated-idents.stderr:34]: I think we do this already, but it's intentionally disabled for UI tests.

File: compiler/rustc_builtin_macros/src/asm.rs (hunk 1 above)
Comment (petrochenkov) [reply compiler/rustc_builtin_macros/src/asm.rs]: The set of expected tokens is not passed as an argument here, it's passed through `p.expected_tokens` and is built automatically by `eat_keyword` and friends.

File: src/test/ui/asm/interpolated-idents.stderr (hunk 2 above)
Comment (petrochenkov) [reply src/test/ui/asm/interpolated-idents.stderr:34]: @Aaron1011 
No, with `-Z deduplicate-diagnostics=yes` the output looks the same.
*Full* diagnostics (error + all the attached notes and labels) are deduplicated, not individual fragments.

**Comment by pietroalbini** [comment]
@bors retry

Yielding priority to the stable release.

**Pull request merged**